### ISSUE TITLE
Magisk cannot detect key events correctly

### ISSUE THREAD

**Issue by pndwal** [topjohnwu/Magisk#4624]
Device: Redmi Note 8T
Android version: 10
Magisk version name: f822ca5
Magisk version code: 23001

**Key Combo Timing for Safe Mode on MIUI / Xiaomi**

I thought I'd try safe mode on my (Xiaomi) devices... (for my own safety).

I have Mi Pad 4 (LTE) w/ LOS18.1 (Android 11), and Redmi Note 8T w/ stock MIUI (Android 10).

[I have not been able to access Safe Mode from RN8T / MIUI from **booted system** any which way 🤨, but no problem on Mi Pad 4 / LOS.]

From **device powered down** I also had issues with timing on both Mi Pad 4 / LOS and RN8T / MIUI.

Holding the pwr button down for more than ~6 seconds after MI / REDMI logos appear causes black screen and fresh reboot, so it is necessary to release pwr after about 4 seconds.

[Releasing pwr then pressing vol- immediately after MI / REDMI logos appear doesn't trigger Safe Mode.]

After releasing pwr at ~ 4 seconds, then waiting (about 10 seconds more) for boot animation (Lineage / MIUI animated logos), then holding vol- until booted, I get Safe Mode but _Magisk has not detected button combo_, and MagiskHide / modules are still enabled.

**Working method:**

However, by holding the pwr button down and releasing it after about 4 seconds, then holding vol- during these logos as well as boot animation logos until booted I get safe mode and Magisk detects / disables MagiskHide and modules.

So clearly Magisk's _button detection is not reliable_ and device may boot to Safe Mode without Magisk detecting it.

It seems there is a critical timing sweet spot, at least for Xiaomi, where Safe Mode is triggered and Magisk also detects the key combination.

**Comment by osm0sis** [comment]
May already be fixed upstream from the last Canary? https://github.com/topjohnwu/Magisk/commit/117d1ed0801c0b9fa144d195c0a89cabc488d176

**Comment by pndwal** [comment]
> May already be fixed upstream from the last Canary? [117d1ed](https://github.com/topjohnwu/Magisk/commit/117d1ed0801c0b9fa144d195c0a89cabc488d176)

No, that must be something else - Entering Safe Mode is not the real issue; Magisk not detecting that this has occurred is.

I'm actually using @vvb2060's latest Alpha w/ TJW commits to 30 August already merged and this remains the same.

This issue can be resolved with some experiments with timing but is causing users to assume Safe Mode method for disabling modules is not working or that their boot problem is not related to incompatible modules when they erroneously assume that these have been disabled...

In short it is misleading and prolonging the agony for those with problem modules causing bootloop.

Would be good if Magisk could piggyback off native Safe mode key detection (probably impossible!) or otherwise properly detect Safe mode once it's booted in order to **always** disable modules...

**Comment by pojavluncher2021** [comment]
> May already be fixed upstream from the last Canary? https://github.com/topjohnwu/Magisk/commit/117d1ed0801c0b9fa144d195c0a89cabc488d176

I suggest making a commit about supporting F1VM based VMs, Plus, Remove magisk, we don't need it since we can use a VM instead,

**Comment by pndwal** [comment]
> > May already be fixed upstream from the last Canary? [117d1ed](https://github.com/topjohnwu/Magisk/commit/117d1ed0801c0b9fa144d195c0a89cabc488d176)
> 
> I suggest making a commit about supporting F1VM based VMs, Plus, Remove magisk, we don't need it since we can use a VM instead,

Sorry, I don't know what you're talking about... This is a Magisk issue regarding bootloop fix for incompatible Magisk modules...

**Comment by osm0sis** [comment]
@pndwal any improvement with 23014? Also you never provided logs and I asked almost a month ago..

**Comment by pndwal** [comment]
> You should probably provide some logs, hey?

Very sorry I missed this ping!...

I'll try to do this in next two days; please say best logs...

> @pndwal any improvement with 23014? Also you never provided logs and I asked almost a month ago..

Just tried again w/ 23014 on RN8T (and standard Xiaomi method:
1) Turn off our smartphone completely.
2) Turn it on normally and when the MIUI logo appears, press and hold the volume down button until the system boots completely --> Safe Mode
3) Rebooted --> System.)

System rebooted w/ Airplane mode on, widgets disabled etc, however **Magisk is still fully functional; modules enabled as before**, SafetyNet passing etc 😕 .

It is still critical to use the following variations to trigger Magisk detection of Safe Mode:
2) Turn it on normally But **continue to hold pwr button.** Release about 4 secs after seconds after standard MI / REDMI logos appear. Then, **don't wait** for the animated MIUI llogo to appear; **immediately press and hold the volume down button** until the system boots completely --> Safe Mode...

**Comment by pndwal** [comment]
> You should probably provide some logs, hey?

Very sorry I missed this ping!...

I'll try to do this in next two days; please say best logs...

> @pndwal any improvement with 23014? Also you never provided logs and I asked almost a month ago..

Just tried again w/ 23014 on RN8T and standard Xiaomi Safe Mode booting method:
1) Turn off our smartphone completely.
2) Turn it on normally and when the MIUI logo appears, press and hold the volume down button until the system boots completely --> Safe Mode
3) Reboot --> System.)

System rebooted w/ Airplane mode on etc, however **Magisk is still fully functional; modules enabled as before**, SafetyNet passing etc 😕 , so no change.

It is still critical to use variations to trigger Magisk detection of Safe Mode.

After trying this many more times, here is my revised (simplified) working method:

1) Turn off our smartphone completely.
2) Turn it on normally. (Can actually release power button immediately.) BUT **don't wait** for the animated MIUI llogo to appear. Press and hold the volume down button **after waiting at least 2 seconds** until the system boots completely --> Safe Mode.
3) Reboot --> System.

It seems that OS checks button state sometime after animated logo appears, but Magisk daemon checks before this so doesn't detect Safe Mode with standard key combo instructions.

On RN8T standard logo shows for approx 33 seconds before animated logo appears. Mi Pad 4 is much quicker.

My tests show
1) Holding volume down before about 3 seconds may not trigger Safe Mode (especially on Mi Pad 4).
2) Holding volume button down from anywhere between 5 and 25 seconds (on RN8T) boots to Safe Mode with Magisk daemon detection.
3) Holding volume button down for 30 seconds (3 seconds before animated logo appears) boots to Safe Mode but Magisk daemon fails to detect this.

**Useful observation:**

Rebooting from safe mode **before unlocking screen** preserves custom launcher, widgets, and other settings on rebooting to system. Only Airplane mode and Magisk modules + Enforce denylist / MagiskHide need resetting for me.

**Another (potentially useful) anomaly:**

I can also trigger Magisk daemon Safe Mode detection without entering Safe Mode at all! (Tested on Mi Pad 4.) Simply:
1) Turn off our smartphone completely.
2) Turn it on normally, but don't wait for the animated MIUI llogo to appear. Press and hold the volume down button after waiting at least 2 seconds and **release it as soon as animated logo appears.** Reboots --> System.

After this no Safe Mode changes occur (airplane mode is still off, etc) but Magisk modules + Enforce denylist / MagiskHide are disabled / need resetting.

Also strangely (perhaps due to Magisk expecting an extra boot) no installed modules appear but they return after a reboot in disabled state, so effectively one extra reboot is required.

hidelist / denylist is populated as soon as Enforce denylist / MagiskHide are toggled on. No reboot needed for that as expected.

It now seems to me that since OS and Magisk daemon Safe Mode detection are clearly independent, John could simply provide his own "Restore/Rescue Magisk Boot Mode" or "Magisk SafeBoot Mode" key combo that simply disables Magisk modules + Enforce denylist / MagiskHide without invoking Safe Mode at all. Effectively (with altered key timing) that's what we already have... Just a few tweaks needed?

👀 PW

**Comment by pndwal** [comment]
> > You should probably provide some logs, hey?
> 
> Very sorry I missed this ping!...
> 
> I'll try to do this in next two days; please say best logs...

As I lack experience diagnosing, I'm not sure what type of issue this is.
**"Bug Reports
For installation issues, upload both boot image and install logs.
For Magisk issues, upload boot logcat or dmesg.
For Magisk app crashes, record and upload the logcat when the crash occurs."**
I'd like to send best logs. Please give a pointer. 🤪

**Comment by osm0sis** [comment]
For something that early dmesg, but logcat could also be useful for all of the safe mode triggering scenarios you've described.

**Comment by pndwal** [comment]
@Osm0sis, sorry I haven't supplied logs as yet... @Didgeridoohan has confirmed I will need to get a logcat or dmesg when booting into Safe Mode using PC / adb and given me a good guide.

Problem has been getting an hour or two to re-do tests with different timing / PC connected etc; my wife and I were both in hospital last week; plenty of time with mobile phone in waiting rooms, drip in arm etc, but not with PC. I will get onto this as soon as I can reasonably give it needed attention...

Further, kmsg/last_kmsg are 0 bytes and there's no console-ramoops in /sys/fs/pstore folder on my device.

FWIW, it seems most likely that this quirk is common to many Xiaomi devices, so anyone with Xiaomi can also test; likely not unique to Xiaomi either FTM...

**Comment by osm0sis** [comment]
Take your time, family comes first. last_kmsg is from previous boot so would be more complicated for you than just getting dmesg output. To get dmesg just running `dmesg > /sdcard/dmesg.log` as root is fine, no PC required, but again, take your time.

**Comment by yujincheng08** [comment]
As far as I tested, my MIUI can trigger safe mode.

**Comment by pndwal** [comment]
> As far as I tested, my MIUI can trigger safe mode.

From booted system? That's interesting... but real issue is with key combo timing for safe mode from ### device powered down on both my Xiaomi devices.

On both devices I can boot to Safe Mode (by holding vol- as soon as animated logo appears) _without_ Magisk detecting the button combo, and MagiskHide / modules remain enabled after reboot as described.

By varying the timing however (ie. by releasing pwr button, then holding vol- _well before_ animated logo until booted) I can boot to safe mode with Magisk detecting disabling MagiskHide and modules.

The problem is that most guides for safe mode on Xiaomi say 'when the MIUI logo appears, press and hold the volume down button until the system boots' or similar. This means that while safe mode detection by Magisk daemon can work with experimentation, users have assumed either that Safe Mode method for disabling modules is not working, or that their boot problem is not related to incompatible modules when the reboot believing they that these have been disabled...

**Comment by pndwal** [comment]
Leaving Canterbury...

**Comment by pndwal** [comment]
Coffee at 7.24...

**Comment by pndwal** [comment]
**Additional Issue with same failure to trigger Magisk Safe Mode on newer devices**

I'm reading reports that late MIUI devices (Mi 11 etc) no longer directly boot to Safe Mode using old key combo. Users report the need to boot to Recovery mode, then select Safe Mode from the menu...

This means that Magisk cannot detect the standard Safe Mode key combo and modules are not disabled even when Safe Mode has been activated, so users cannot use Safe Mode method to restore boot when modules are incompatible.

**Comment by osm0sis** [comment]
Doesn't it also trigger Magisk through the safe mode prop being set? What prop is MIUI using? Maybe this part can be resolved that way.

**Comment by pndwal** [comment]
> Doesn't it also trigger Magisk through the safe mode prop being set? What prop is MIUI using? Maybe this part can be resolved that way.

Not sure about props here, but last comment may be a false alarm in fact, sorry... One user who said direct key combo not working in MIUI 13 just said "Sorry..., but it is possible to start safe mode by "long press the power, the Mi LOGO appears and vibrate to let go, and then keep pressing the volume - until the power is turned on"."... Other users report key combo not working with custom ROMs however... Il edit post above. 😬

**Comment by pndwal** [comment]
I can confirm that the issue was NOT fixed...

Sorry i'd missed this request... I did put some comments re. Shana's fix and why I was also sceptical it would work:

> [snip]
>
> AS I mentioned, this is guesswork on my part, and it does seem that there may only be nanoseconds between the setting of these props, so I'm actually equally sceptical that this will be a proper fix... It still looks to me as if Magisks critical bootstage checks may simply have timed out by the time the animated logo appears on some devices while Android system seems to set the props early but continue checking vol- until animated logo and revert settings unless vol- is still depressed at that point...
>
> So we shall see... I'm happy to see at least some early work on this...

https://forum.xda-developers.com/t/magisk-general-support-discussion.3432382/post-87901089

Anyway, I've just tested this again on Mi Pad 4 w/ 25206 and see the same behaviour... 

Safe mode booted by powering up from off state then holding vol- when animated logo appears results in Magisk modules still enabled after a reboot...

As previously,  Safe mode booted by powering up from off state then holding vol- _well before_ animated logo appears results in Magisk modules still enabled after a reboot...

... Many thanks for your efforts/attention to this...

**Comment by yujincheng08** [comment]
@pndwal how about booting into recovery and then select safe mode and reboot into normal mode?

**Comment by pndwal** [comment]
> @pndwal, how about booting into recovery, selecting safe mode, and rebooting into normal mode?

Sadly it can't work on either of my devices...

Neither Mi Pad 4 w/ LOS 18.1 & TWRP 3.5.0_9-0 nor Redmi Note 8T w/ stock MIUI (but still Android 10) offer a Reboot to Safe mode option in respective recoveries...

I guess that may be a workable solution for some/later devices/OSs...

-----------
The issue caused by current boot to Safe mode behaviour may be resolved by either fixing Magisk demons failure to properly respond to key-press data or simply by explaining to users that entering Safe mode is no guarantee that Magisk has detected this / modules are disabled and that they may need to alter key-press timing and try again...

Without this information, as I said above:

> The problem is... that while safe mode detection by Magisk daemon can work with experimentation, users have assumed either that Safe Mode method for disabling modules is not working, or that their boot problem is not related to incompatible modules when they reboot believing that these have been disabled...

... Many thanks for your efforts / revisiting this...

**Comment by vvb2060** [comment]
This is not possible, magisk must disable modules before system service starts.  Therefore the key press must be detected earlier than system.

**Comment by pndwal** [comment]
> This is not possible, magisk must disable modules before system service starts. Therefore the key press must be detected earlier than system.

Sorry, is neither I suggested alternative possible?... Can't we 'simply **explain to users that entering Safe mode is no guarantee that Magisk has detected this / modules are disabled**, and **that they may need to alter key-press timing and try again** so that users don't assume either that Safe Mode method for disabling modules does not work, or that their boot problem is not related to incompatible modules when they reboot because of believing that these have actually been disabled'?

... It does appear that you're saying exactly what I said. but the information should be added to the first TJW FAQ Q. etc...

**Comment by pndwal** [comment]
As I noted above, 

> By varying the timing however (ie. by releasing pwr button, then holding vol- _well before_ animated logo until booted) I can boot to safe mode with Magisk detecting disabling MagiskHide and modules.

but

> The problem is that most guides for safe mode on Xiaomi say '_when the MIUI logo appears_, press and hold the volume down button until the system boots' or similar

and this method fails... Also,

> This means that... users have assumed either that Safe Mode method for disabling modules is not working, or that their boot problem is not related to incompatible modules when they reboot believing that these have been disabled...
